File: python/tav2/adapters/renpy/backfill.py
"""把译文回填到 tl/<语言>/*.rpy。

移植自 TranslateAgent v1 的 translate_agent/backfill.py。
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

from tav2.adapters.renpy.tlparser import parse_tl_directory, rebuild_chunk


def backfill_machine(
    game_dir: str | Path,
    lang: str,
    dialogue_map: dict[tuple[str, str], dict[int, str]],
    string_map: dict[tuple[str, str], str],
) -> dict[str, int]:
    """直写模式：按 (文件名, 标识符) -> {行号: 译文} 与 (文件名, old) -> 译文 回填 tl。
    返回统计 {applied, skipped, unchanged}。"""

    return _apply_maps(game_dir, lang, dialogue_map, string_map)


def _apply_maps(
    game_dir: str | Path,
    lang: str,
    dialogue_map: dict[tuple[str, str], dict[int, str]],
    string_map: dict[tuple[str, str], str],
) -> dict[str, int]:
    """核心回填：把映射应用到 tl/<lang>/*.rpy，跳过无变化的文件。"""

    game_dir = Path(game_dir)
    gamedir = game_dir / "game" if (game_dir / "game").is_dir() else game_dir
    tl_dir = gamedir / "tl" / lang
    files = parse_tl_directory(game_dir, lang)
    stats = {"applied": 0, "skipped": 0, "unchanged": 0}

    for path, chunks in files:
        rel = path.relative_to(tl_dir).as_posix()
        changed = False
        new_chunks: list[list[str]] = []
        file_string_map = {
            old: translation
            for (filename, old), translation in string_map.items()
            if filename == rel
        }
        for chunk in chunks:
            say_translations: dict[int, str] | None = None
            if chunk.kind == "dialogue":
                per_say = dialogue_map.get((rel, chunk.identifier or ""))
                if per_say:
                    say_translations = per_say
            rebuilt = rebuild_chunk(chunk, say_translations, file_string_map)
            if rebuilt != chunk.raw:
                changed = True
            new_chunks.append(rebuilt)

        if not changed:
            continue
        text = path.read_text(encoding="utf-8", errors="replace")
        trailing = "\n" if text.endswith("\n") else ""
        body = "\n".join(line for chunk in new_chunks for line in chunk)
        path.write_text(body + trailing, encoding="utf-8")
        stats["applied"] += 1

    return stats


APPLY_STATUSES = {"已确认", "已修改"}


def iter_applied_rows(rows: list[dict[str, Any]], force: bool = False):
    """按审校状态过滤本次应回填的行（跳过/无译文/未确认除外）。"""

    for row in rows:
        status = str(row.get("状态") or "待审")
        if status == "跳过" or (status not in APPLY_STATUSES and not force):
            continue
        translation = str(row.get("人工译文") or row.get("机器译文") or "").strip()
        if not translation:
            continue
        yield row


def backfill_review(
    game_dir: str | Path,
    lang: str,
    rows: list[dict[str, Any]],
    force: bool = False,
) -> dict[str, int]:
    """回填审校行到 tl 文件（可选审校模式的回填入口）。返回统计（含 skip_reasons）。"""

    dialogue_map: dict[tuple[str, str], dict[int, str]] = {}
    string_map: dict[tuple[str, str], str] = {}
    skipped = 0
    skip_reasons: dict[str, int] = {}

    def skip(reason: str) -> None:
        nonlocal skipped
        skipped += 1
        skip_reasons[reason] = skip_reasons.get(reason, 0) + 1

    for row in rows:
        status = str(row.get("状态") or "待审")
        if status == "跳过":
            skip("状态=跳过")
            continue
        if status not in APPLY_STATUSES and not force:
            skip(f"状态={status}（需 已确认/已修改，或 --force）")
            continue
        translation = str(row.get("人工译文") or row.get("机器译文") or "").strip()
        if not translation:
            skip("人工译文与机器译文均为空")
            continue
        row_type = str(row.get("类型") or "")
        filename = str(row.get("文件") or "")
        if row_type == "dialogue":
            identifier = str(row.get("标识符") or "")
            try:
                say_index = int(row.get("序号") or 0)
            except (TypeError, ValueError):
                say_index = 0
            dialogue_map.setdefault((filename, identifier), {})[say_index] = translation
        elif row_type == "string":
            old = str(row.get("标识符") or "")
            string_map[(filename, old)] = translation
        else:
            skip(f"未知类型={row_type or '（空）'}")

    stats = _apply_maps(game_dir, lang, dialogue_map, string_map)
    stats["skipped"] += skipped
    stats["skip_reasons"] = skip_reasons  # type: ignore[assignment]
    return stats

File: src/engine/adapters/renpy/backfill.ts
/**
 * 把译文回填到 tl/<语言>/*.rpy。
 * 移植自 tav2 的 adapters/renpy/backfill.py（backfill_machine）。
 */
import { readFileSync, writeFileSync } from 'node:fs'
import { relative, sep } from 'node:path'
import { parseTlDirectory, rebuildChunk, tlRoot } from './tlparser'

/**
 * dialogue_map 的结构：键为 `${filename}|${identifier}`，值为 say_index -> 译文。
 * string_map 的结构：键为 `${filename}|${old}`，值为译文。
 * 既接受 Map 也接受嵌套 Record/扁平 Record。
 */
export type DialogueMapInput =
  | Map<string, Map<string | number, string>>
  | Record<string, Record<string | number, string>>

export type StringMapInput = Map<string, string> | Record<string, string>

export interface BackfillStats {
  applied: number
  skipped: number
  unchanged: number
}

/** 规范化 dialogue_map 为 (filename, identifier, say_index -> 译文) 列表。 */
function normalizeDialogue(input: DialogueMapInput): Array<[string, string, Map<number, string>]> {
  const out: Array<[string, string, Map<number, string>]> = []
  if (input instanceof Map) {
    for (const [key, inner] of input) {
      const [filename, identifier] = splitKey(key)
      const perSay = new Map<number, string>()
      for (const [k, v] of inner) perSay.set(Number(k), v)
      out.push([filename, identifier, perSay])
    }
    return out
  }
  for (const [key, inner] of Object.entries(input)) {
    const [filename, identifier] = splitKey(key)
    const perSay = new Map<number, string>()
    for (const [k, v] of Object.entries(inner)) perSay.set(Number(k), String(v))
    out.push([filename, identifier, perSay])
  }
  return out
}

/** 规范化 string_map 为 (filename, old, translation) 列表。 */
function normalizeStrings(input: StringMapInput): Array<[string, string, string]> {
  const out: Array<[string, string, string]> = []
  if (input instanceof Map) {
    for (const [key, translation] of input) {
      const [filename, old] = splitKey(key)
      out.push([filename, old, translation])
    }
    return out
  }
  for (const [key, translation] of Object.entries(input)) {
    const [filename, old] = splitKey(key)
    out.push([filename, old, String(translation)])
  }
  return out
}

function splitKey(key: string): [string, string] {
  const idx = key.indexOf('|')
  if (idx < 0) return [key, '']
  return [key.slice(0, idx), key.slice(idx + 1)]
}

function toPosixRel(file: string, root: string): string {
  return relative(root, file).split(sep).join('/')
}

/**
 * 直写模式：按 (文件名, 标识符) -> {行号: 译文} 与 (文件名, old) -> 译文 回填 tl。
 * 返回统计 {applied, skipped, unchanged}（与 Python 基线一致：applied=写入文件数，skipped/unchanged 恒 0）。
 */
export function backfillMachine(
  gameDir: string,
  lang: string,
  dialogueMap: DialogueMapInput,
  stringMap: StringMapInput,
): BackfillStats {
  const normalizedDialogue = normalizeDialogue(dialogueMap)
  const normalizedStrings = normalizeStrings(stringMap)

  const dialogueByFile = new Map<string, Map<string, Map<number, string>>>()
  for (const [filename, identifier, perSay] of normalizedDialogue) {
    if (!dialogueByFile.has(filename)) dialogueByFile.set(filename, new Map())
    dialogueByFile.get(filename)!.set(identifier, perSay)
  }

  const files = parseTlDirectory(gameDir, lang)
  const stats: BackfillStats = { applied: 0, skipped: 0, unchanged: 0 }
  const root = tlRoot(gameDir, lang)

  for (const [path, chunks] of files) {
    const rel = toPosixRel(path, root)
    const perFileDialogue = dialogueByFile.get(rel)
    const perFileStrings = normalizedStrings
      .filter(([filename]) => filename === rel)
      .map(([, old, translation]) => [old, translation] as [string, string])

    let changed = false
    const newChunks: string[][] = []
    for (const chunk of chunks) {
      let sayTranslations: Map<number, string> | undefined
      if (chunk.kind === 'dialogue') {
        const perSay = perFileDialogue?.get(chunk.identifier ?? '')
        if (perSay && perSay.size > 0) sayTranslations = perSay
      }
      const stringMapForFile =
        perFileStrings.length > 0 ? new Map(perFileStrings) : undefined
      const rebuilt = rebuildChunk(chunk, sayTranslations, stringMapForFile)
      if (rebuilt.join('\n') !== chunk.raw.join('\n')) changed = true
      newChunks.push(rebuilt)
    }

    if (!changed) continue
    const text = readFileSync(path, 'utf8')
    const trailing = text.endsWith('\n') ? '\n' : ''
    const body = newChunks.map((chunkLines) => chunkLines.join('\n')).join('\n')
    writeFileSync(path, body + trailing, 'utf8')
    stats.applied += 1
  }

  return stats
}
